HealthVault/Classes/Caching/MHVThingCacheDatabase.m

```
@@ -0,0 +1,570 @@
+}
+
+#pragma mark - Create Objects
+
+- (MHVCachedThing *)newThingForRecord:(MHVCachedRecord *)record
+{
+    if (self.deleted || !record)
+    {
+        return nil;
+    }
+    
+    __block MHVCachedThing *thing;
+    
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         thing = [NSEntityDescription insertNewObjectForEntityForName:@"MHVCachedThing"
+                                               inManagedObjectContext:self.managedObjectContext];
+         thing.record = record;
+     }];
+    
+    return thing;
+}
+
+- (MHVCachedRecord *)newRecord:(NSString *)recordId
+{
+    if (self.deleted || !recordId)
+    {
+        return nil;
+    }
+    
+    __block MHVCachedRecord *record;
+    
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         record = [NSEntityDescription insertNewObjectForEntityForName:@"MHVCachedRecord"
+                                                inManagedObjectContext:self.managedObjectContext];
+         record.recordId = recordId;
+         record.lastOperationSequenceNumber = 0;
+         record.lastSyncDate = [NSDate dateWithTimeIntervalSince1970:0];
+         record.isValid = YES;
+         
+         [self saveContext];
+     }];
+    
+    return record;
+}
+
+#pragma mark - Delete
+
+- (void)deleteRecord:(NSString *)recordId
+{
+    MHVASSERT_PARAMETER(recordId);
+    
+    if (self.deleted || !recordId)
+    {
+        return;
+    }
+    
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         MHVCachedRecord *record = (MHVCachedRecord *)[self fetchCachedRecord:recordId];
+         if (record)
+         {
+             [self.managedObjectContext deleteObject:record];
+             [self saveContext];
+         }
+     }];
+}
+
+- (NSError *_Nullable)deleteThingIds:(NSArray<NSString *> *)thingIds recordId:(NSString *)recordId
+{
+    MHVASSERT_PARAMETER(thingIds);
+    MHVASSERT_PARAMETER(recordId);
+    
+    if (self.deleted || !thingIds || thingIds.count == 0)
```

Need to be synchronized - Not thread safe. Also, it would appear that the thingIds parameter is required (and required that there be at least one element) - we should return an error here.